bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/auth/client/oauth2/AccessTokenResponse.java

```diff
@@ -0,0 +1,223 @@
+     *            This causes the OAuthToken to expire earlier then the stated expiry-time given
+     *            by the authorization server.
+     *
+     * @return true if object is newly created, or expired. Otherwise false.
+     */
+    public boolean isExpired(@NonNull Date givenTime, int tokenExpiresInBuffer) {
+        return (created == null || created.getTime() + TimeUnit.SECONDS.toMillis(expiresIn)
+                - TimeUnit.SECONDS.toMillis(tokenExpiresInBuffer) <= givenTime.getTime());
+    }
+
+    public String getAccessToken() {
+        return accessToken;
+    }
+
+    public void setAccessToken(String accessToken) {
+        this.accessToken = accessToken;
+    }
+
+    public String getTokenType() {
+        return tokenType;
+    }
+
+    public void setTokenType(String tokenType) {
+        this.tokenType = tokenType;
+    }
+
+    public long getExpiresIn() {
+        return expiresIn;
+    }
+
+    public void setExpiresIn(long expiresIn) {
+        this.expiresIn = expiresIn;
+    }
+
+    public String getRefreshToken() {
+        return refreshToken;
+    }
+
+    public void setRefreshToken(String refreshToken) {
+        this.refreshToken = refreshToken;
+    }
+
+    public String getScope() {
+        return scope;
+    }
+
+    public void setScope(String scope) {
+        this.scope = scope;
+    }
+
+    public String getState() {
+        return state;
+    }
+
+    public void setState(String state) {
+        this.state = state;
+    }
+
+    public Date getCreated() {
+        return created;
+    }
+
+    public void setCreated(Date created) {
+        this.created = created;
+    }
+
+    @Override
+    public Object clone() {
+        try {
+            return super.clone();
+        } catch (CloneNotSupportedException e) {
+            return null; // not possible
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(accessToken, refreshToken, tokenType, scope, expiresIn, created);
+    }
+
+    @Override
+    public boolean equals(Object thatAuthTokenObj) {
+
+        if (this == thatAuthTokenObj) {
+            return true;
+        }
+
+        // Exact match since class is final
+        if (thatAuthTokenObj == null || !this.getClass().equals(thatAuthTokenObj.getClass())) {
+            return false;
+        }
+
+        AccessTokenResponse that = (AccessTokenResponse) thatAuthTokenObj;
+
+        return Objects.equals(this.accessToken, that.accessToken) && Objects.equals(this.tokenType, that.tokenType)
+                && Objects.equals(this.expiresIn, that.expiresIn)
+                && Objects.equals(this.refreshToken, that.refreshToken) && Objects.equals(this.scope, that.scope)
+                && Objects.equals(this.state, that.state) && Objects.equals(this.created, that.created);
+
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
```

Comment:
You could also use -> Objects.toStringHelper() from guava

Reply:
We already discussed about this but I feel the manual intervention to manage ":" and spaces in the output string could have been done in a better way using `StringJoiner` from Java 8 or as suggested previously using Guava. If you don't have any reference to Guava library, don't use it but if you have it, you could write the following:

```java
      Objects.toStringHelper(this)
             .add("AccessToken", accessToken)
             .add("TokenType", tokenType)
             .add("RefreshToken", refreshToken)
             .add("Scope", scope)
             .add("ExpiresIn", expiresIn)
             .add("State", state)
             .add("Created", created)
             .toString();
```
It would manage everything for you and the manual complexities are abstracted out.

Reply:
Replaced with Eclipse auto-generated.